gax-java/gax/src/test/java/com/google/api/gax/tracing/MetricsTracerTest.java

```diff
@@ -0,0 +1,247 @@
+    metricsTracer.attemptCancelled();
+    metricsTracer.operationCancelled();
+
+    long count = 1;
+    Map<String, String> attributes =
+        ImmutableMap.of(
+            "status", "CANCELLED",
+            "method_name", "fake_service.fake_method");
+
+    verify(metricsRecorder).recordAttemptCount(count, attributes);
+    verify(metricsRecorder).recordOperationCount(count, attributes);
+    verify(metricsRecorder).recordAttemptLatency(anyDouble(), eq(attributes));
+    verify(metricsRecorder).recordOperationLatency(anyDouble(), eq(attributes));
+
+    verifyNoMoreInteractions(metricsRecorder);
+  }
+
+  @Test
+  public void testAttemptFailedRetriesExhaustedExample() {
+    // initialize mock-request
+    Object mockRequest = new Object();
+
+    // Attempt #1
+    metricsTracer.attemptStarted(mockRequest, 0);
+    ApiException error0 =
+        new DeadlineExceededException(
+            "deadline exceeded", null, new FakeStatusCode(Code.DEADLINE_EXCEEDED), false);
+
+    metricsTracer.attemptFailedRetriesExhausted(error0);
+    metricsTracer.operationFailed(error0);
+
+    long count = 1;
+    Map<String, String> attributes =
+        ImmutableMap.of(
+            "status", "DEADLINE_EXCEEDED",
+            "method_name", "fake_service.fake_method");
+
+    verify(metricsRecorder).recordAttemptCount(count, attributes);
+    verify(metricsRecorder).recordOperationCount(count, attributes);
+    verify(metricsRecorder).recordAttemptLatency(anyDouble(), eq(attributes));
+    verify(metricsRecorder).recordOperationLatency(anyDouble(), eq(attributes));
+
+    verifyNoMoreInteractions(metricsRecorder);
+  }
+
+  @Test
+  public void testAttemptPermanentFailureExample() {
+    // initialize mock-request
+    Object mockRequest = new Object();
+
+    // Attempt #1
+    metricsTracer.attemptStarted(mockRequest, 0);
+
+    ApiException error0 =
+        new NotFoundException("not found", null, new FakeStatusCode(Code.NOT_FOUND), false);
+
+    metricsTracer.attemptFailedRetriesExhausted(error0);
+    metricsTracer.operationFailed(error0);
+
+    long count = 1;
+    Map<String, String> attributes =
+        ImmutableMap.of(
+            "status", "NOT_FOUND",
+            "method_name", "fake_service.fake_method");
+
+    verify(metricsRecorder).recordAttemptCount(count, attributes);
+    verify(metricsRecorder).recordOperationCount(count, attributes);
+    verify(metricsRecorder).recordAttemptLatency(anyDouble(), eq(attributes));
+    verify(metricsRecorder).recordOperationLatency(anyDouble(), eq(attributes));
+
+    verifyNoMoreInteractions(metricsRecorder);
+  }
+
+  @Test
+  public void testErrorConversion() {
+    for (Code code : Code.values()) {
+      ApiException error = new ApiException("fake error", null, new FakeStatusCode(code), false);
+      String errorCode = metricsTracer.extractStatus(error);
+      assertThat(errorCode).isEqualTo(code.toString());
+    }
+  }
+
+  // this test needs detailed attention
```

Comment:
@blakeli0 @lqiu96 Can you guys confirm if this a valid scenario ?

Reply:
I'm not sure we need this scenario. Can you try to re-organize this test file as I suggested in other comments first?

Reply:
I've removed this test as it is testing multiple methods in the same test.